### PR TITLE
Fix each bean with a bean annotated `@Named` and `@Primary`

### PR DESCRIPTION
**Pull request by dstepanov**

### CODE REVIEW

**Review by timyates — APPROVED** [review]
This fixes the initialization error we were seeing with Kafka-streams over here

https://github.com/micronaut-projects/micronaut-kafka/pull/641

👍

**Review by graemerocher — APPROVED** [review]

File: inject-java/src/test/groovy/io/micronaut/inject/configproperties/eachbeanparameter/DefaultDataSource.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import io.micronaut.context.annotation.Primary;
+import io.micronaut.context.annotation.Requires;
+import jakarta.inject.Named;
+import jakarta.inject.Singleton;
+
+@Singleton
+@Requires(property = "spec", value = "EachBeanParameterSpec")
+@Named("default")
+@Primary
+class DefaultDataSource extends AbstractDataSource {
```
Comment (sdelamo) [inline inject-java/src/test/groovy/io/micronaut/inject/configproperties/eachbeanparameter/DefaultDataSource.java:12]: test passes even if I delete this file.

Comment (sdelamo) [reply inject-java/src/test/groovy/io/micronaut/inject/configproperties/eachbeanparameter/DefaultDataSource.java:12]: I added a test which verifies there are four beans of type `AbstractDataSource`.  https://github.com/micronaut-projects/micronaut-core/pull/8510/commits/d0311b6a275af8637ab419445142151b347a7ef3 

I expected to be only three. With `DefaultDataSource` taking precedence over the bean created via `@EachProperty`.

File: inject-java/src/test/groovy/io/micronaut/inject/configproperties/eachbeanparameter/EachBeanParameterSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package io.micronaut.inject.configproperties.eachbeanparameter
+
+import io.micronaut.annotation.processing.test.AbstractTypeElementSpec
+import io.micronaut.context.ApplicationContext
+
+class EachBeanParameterSpec extends AbstractTypeElementSpec {
+
+    void 'test interceptor on an event'() {
```
Comment (sdelamo) [inline inject-java/src/test/groovy/io/micronaut/inject/configproperties/eachbeanparameter/EachBeanParameterSpec.groovy]: can we provide a better test name?

**Review by sdelamo — CHANGES_REQUESTED** [review]

File: inject-java/src/test/groovy/io/micronaut/inject/configproperties/eachbeanparameter/DefaultDataSource.java (hunk 1 above)
Comment (sdelamo) [reply inject-java/src/test/groovy/io/micronaut/inject/configproperties/eachbeanparameter/DefaultDataSource.java:12]: add an extra verification. https://github.com/micronaut-projects/micronaut-core/pull/8510/commits/06747f30d0987b3ded1eaec833bc7a9552c816fd

File: inject-java/src/test/groovy/io/micronaut/inject/configproperties/eachbeanparameter/MyFactory.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+@Factory
+public class MyFactory {
+
+    @EachBean(AbstractDataSource.class)
+    MyHelper buildHelper() {
+        return new MyHelper();
+    }
+
+    @Context // The context should load without properties and properly fill the parameter from DefaultDataSource
```
Comment (sdelamo) [inline inject-java/src/test/groovy/io/micronaut/inject/configproperties/eachbeanparameter/MyFactory.java:16]: test passes whether `@Context` is defined or not. Should we remove it?

Comment (dstepanov) [reply inject-java/src/test/groovy/io/micronaut/inject/configproperties/eachbeanparameter/MyFactory.java:16]: We should keep, it's was causing the problem

Comment (sdelamo) [reply inject-java/src/test/groovy/io/micronaut/inject/configproperties/eachbeanparameter/MyFactory.java:16]: With your code changes, it seems unnecessary but without the changes to  `BeanDefinitionDelegate` only one bean of type `MyBean` was registered.

**Pull request merged**